Load testing document (#593)
vdaas/vald: docs/performance/loadtest.md

# Load Testing

This document describes how to perform the load testing in Vald cluster or Vald agent using Vald Load Testing Tools.

## Overview

> We develop our own Load Testing Tools to perform the load testing in Vald cluster or Vald agent. You can perform the load test on Vald to test the insert and search performance in Vald.

Reviewer: ```suggestion
We develop our own Load Testing Tools to perform the load testing in Vald cluster or Vald agent.
You can perform the load test on Vald to test the insert and search performance in Vald.
```

## Prerequisite

- Vald Cluster or Vald Agent

    Please refer to [here](https://vald.vdaas.org/docs/tutorial/get-started/) for the installation guide of Vald Cluster or Vald Agent.

> - Docker

Reviewer: Could you show the minimum version?

Author: Sorry I am not sure the minimum version of docker to execute the load testing tools... 
I think it works on most version of docker though.... 
What do you think?

Reviewer: I think you show the version of docker which you run the load test.

## Install load test tools

> We provide a docker image for the load testing. Please download the image using the below command.

Reviewer: ```suggestion
We provide a docker image for the load testing.
Please download the image using the below command.
```

```bash
docker pull vdaas/vald-loadtest
```

## Configure load test tools

Please refer to the [Sample configuration file](https://github.com/vdaas/vald/blob/master/cmd/tools/cli/loadtest/sample.yaml) to configure the Vald Load Testing Tools.

> Here is the important configurations and following the explaination.

Reviewer 2: ```suggestion
Here is the important configurations and following the explanation.
```

> | Name        | Description                                                                                                                                                      | Example                                             |

Reviewer: the table will make us to understand clearly in Github, but it may not be shown all of the below table in the public document framework

Author: I think it works maybe.
https://raw.githubusercontent.com/vdaas/vald/master/CHANGELOG.md

Reviewer: I meant whether work or not.
I wonder whether the whole of table will be shown without scrollbar or not.
e.g. https://vald.vdaas.org/docs/contributing/coding-style/#table-driven-test

| dataset     | The dataset used in `insert` and `streaminsert` operation.                                                                                                       |                                                     |
| concurrency | The number of concurrent execution of the load test.                                                                                                             |                                                     |
| batch_size  | The batch size of the dataset.                                                                                                                                   |                                                     |
| addr        | The cluster you want to test                                                                                                                                     |                                                     |

## Execute load test

Download the sample configuration file.

```bash
wget https://raw.githubusercontent.com/vdaas/vald/master/cmd/tools/cli/loadtest/sample.yaml
mv sample.yaml config.yaml
```

> Please refer to the above section to configure the load testing tool.

Reviewer: What is the above section?
It would be better to show the concrete name instead of `the above section`.

Author: What do you think about this?
```suggestion
Please refer to the [configure load test tools section](#configure-load-test-tools) to configure the load testing tool.
```

2020-07-13 17:10:55	[INFO]:	progress 370 requests, 6166.631312[vps], error: 0
2020-07-13 17:10:58	[INFO]:	progress 558 requests, 6199.932718[vps], error: 0
2020-07-13 17:11:01	[INFO]:	progress 727 requests, 6058.303243[vps], error: 0
2020-07-13 17:11:04	[INFO]:	progress 910 requests, 6066.658704[vps], error: 0
2020-07-13 17:11:05	[INFO]:	result:Gateway	32	100	6060.743248
2020-07-13 17:11:05	[WARN]:	terminated signal received daemon will stopping soon...
2020-07-13 17:11:05	[INFO]:	executing daemon pre-stop function
2020-07-13 17:11:05	[INFO]:	executing daemon stop function
2020-07-13 17:11:05	[INFO]:	executing daemon post-stop function
2020-07-13 17:11:05	[ERR]:	error occured in runner.Wait at load test: context canceled
2020-07-13 17:11:05	[WARN]:	daemon stopped
```

These lines show the result of the load test.

```bash
2020-07-13 08:10:45 [DEBG]: start loading: random-786-100000
...
2020-07-13 17:10:49 [INFO]: start load test(Gateway, Insert)
...
2020-07-13 17:11:05	[INFO]:	result:Gateway	32	100	6060.743248
```

This means that the `Insert Gateway` mode is used, which means the dataset `random-786-100000` will insert into the Vald Cluster.

> The line `result:Gateway  32  100 6060.743248` means that the Gateway mode is used, with `32` concurrent execution and `100` batch size, with the VPS of `6060.743248`.

Reviewer: ```suggestion
The line `result:Gateway  32  100 6060.743248` means that the Gateway mode is used, with `32` concurrent execution and `100` batch size, with the VPS (vector per sec.) of `6060.743248`.
```

Which means that is performed the insertion of 6060 vectors into the Vald Cluster per second from the Vald Load Testing tools.

> The result includes all the network latency, and the filtering latency.

Reviewer 2: ```suggestion
The result includes all the network latency and filtering latency.
```
